compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1054,10 +1087,74 @@ impl<'hir> LoweringContext<'_, 'hir> {
         body: impl FnOnce(&mut Self) -> hir::Expr<'hir>,
     ) -> hir::BodyId {
         self.lower_body(|this| {
-            (
-                this.arena.alloc_from_iter(decl.inputs.iter().map(|x| this.lower_param(x))),
-                body(this),
-            )
+            let params =
+                this.arena.alloc_from_iter(decl.inputs.iter().map(|x| this.lower_param(x)));
+            let result = body(this);
+
+            let opt_contract = this.contract.take();
+
+            // { body }
+            // ==>
+            // { contract_requires(PRECOND); { body } }
+            let result: hir::Expr<'hir> = if let Some(contract) = opt_contract {
```

Comment:
somewhat tempted to use

```suggestion
            let Some(contract) = opt_contract else { return (params, result) };
```